File: binding/java/test/ai/picovoice/porcupine/PorcupineTest.java
/*
    Copyright 2018-2023 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.porcupine;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.UnsupportedAudioFileException;
import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;


public class PorcupineTest {

    private final String accessKey = System.getProperty("pvTestingAccessKey");
    private Porcupine porcupine;

    private static Stream<Arguments> singleKeywordProvider() throws IOException {
        final JsonObject testDataJson = PorcupineTestUtils.loadTestData();
        final JsonArray singleKeywordData = testDataJson
                .getAsJsonObject("tests")
                .getAsJsonArray("singleKeyword");

        final ArrayList<Arguments> testArgs = new ArrayList<>();
        for (int i = 0; i < singleKeywordData.size(); i++) {
            final JsonObject testData = singleKeywordData.get(i).getAsJsonObject();
            final String language = testData.get("language").getAsString();
            final String keyword = testData.get("wakeword").getAsString();
            final String testAudio = keyword.replace(' ', '_') + ".wav";
            testArgs.add(Arguments.of(
                    language,
                    keyword,
                    testAudio)
            );
        }
        return testArgs.stream();
    }

    private static Stream<Arguments> multipleKeywordsProvider() throws IOException {
        final JsonObject testDataJson = PorcupineTestUtils.loadTestData();
        final JsonArray multipleKeywordData = testDataJson
                .getAsJsonObject("tests")
                .getAsJsonArray("multipleKeyword");

        final Gson gson = new Gson();
        final ArrayList<Arguments> testArgs = new ArrayList<>();
        for (int i = 0; i < multipleKeywordData.size(); i++) {
            final JsonObject testData = multipleKeywordData.get(i).getAsJsonObject();
            final String language = testData.get("language").getAsString();
            final String[] keywords = gson.fromJson(testData.get("wakewords"), String[].class);
            final String testAudio = PorcupineTestUtils.appendLanguage("multiple_keywords", language) + ".wav";
            final ArrayList<Integer> groundTruth = gson.fromJson(
                    testData.get("groundTruth"),
                    new TypeToken<ArrayList<Integer>>() {
                    }.getType());
            testArgs.add(Arguments.of(
                    language,
                    keywords,
                    testAudio,
                    groundTruth)
            );

        }
        return testArgs.stream();
    }

    @AfterEach
    void tearDown() {
        porcupine.delete();
    }

    @Test
    void getVersion() throws PorcupineException {
        porcupine = new Porcupine.Builder()
                .setAccessKey(accessKey)
                .setBuiltInKeyword(Porcupine.BuiltInKeyword.PORCUPINE)
                .build();
        assertTrue(porcupine.getVersion() != null && !porcupine.getVersion().equals(""));
    }

    @Test
    void getFrameLength() throws PorcupineException {
        porcupine = new Porcupine.Builder()
                .setAccessKey(accessKey)
                .setBuiltInKeyword(Porcupine.BuiltInKeyword.PORCUPINE)
                .build();
        assertTrue(porcupine.getFrameLength() > 0);
    }

    @org.junit.jupiter.api.Test
    void getSampleRate() throws PorcupineException {
        porcupine = new Porcupine.Builder()
                .setAccessKey(accessKey)
                .setBuiltInKeyword(Porcupine.BuiltInKeyword.PORCUPINE)
                .build();
        assertTrue(porcupine.getSampleRate() > 0);
    }

    private void runTestCase(String audioFileName, ArrayList<Integer> expectedResults)
            throws IOException, UnsupportedAudioFileException, PorcupineException {
        int frameLen = porcupine.getFrameLength();
        String audioFilePath = PorcupineTestUtils.getTestAudioFilePath(audioFileName);
        File testAudioPath = new File(audioFilePath);

        AudioInputStream audioInputStream = AudioSystem.getAudioInputStream(testAudioPath);
        assertEquals(audioInputStream.getFormat().getFrameRate(), 16000);

        int byteDepth = audioInputStream.getFormat().getFrameSize();
        int bufferSize = frameLen * byteDepth;
        byte[] pcm = new byte[bufferSize];
        short[] porcupineFrame = new short[frameLen];
        int numBytesRead;
        ArrayList<Integer> results = new ArrayList<>();
        while ((numBytesRead = audioInputStream.read(pcm)) != -1) {

            if (numBytesRead / byteDepth == frameLen) {

                ByteBuffer
                        .wrap(pcm)
                        .order(ByteOrder.LITTLE_ENDIAN)
                        .asShortBuffer()
                        .get(porcupineFrame);
                int result = porcupine.process(porcupineFrame);
                assertTrue(result >= -1);
                if (result >= 0) {
                    results.add(result);
                }
            }
        }

        assertEquals(results, expectedResults);
    }

    @Test
    void testBuiltinKeywords()
            throws IOException, UnsupportedAudioFileException, PorcupineException {
        final Porcupine.BuiltInKeyword[] keywords = new Porcupine.BuiltInKeyword[]{
                Porcupine.BuiltInKeyword.ALEXA,
                Porcupine.BuiltInKeyword.AMERICANO,
                Porcupine.BuiltInKeyword.BLUEBERRY,
                Porcupine.BuiltInKeyword.BUMBLEBEE,
                Porcupine.BuiltInKeyword.GRAPEFRUIT,
                Porcupine.BuiltInKeyword.GRASSHOPPER,
                Porcupine.BuiltInKeyword.PICOVOICE,
                Porcupine.BuiltInKeyword.PORCUPINE,
                Porcupine.BuiltInKeyword.TERMINATOR
        };

        porcupine = new Porcupine.Builder()
                .setAccessKey(accessKey)
                .setBuiltInKeywords(keywords)
                .build();

        runTestCase(
                "multiple_keywords.wav",
                new ArrayList<>(Arrays.asList(7, 0, 1, 2, 3, 4, 5, 6, 7, 8)));
    }

    @Test
    void testWithNonAsciiModelName()
            throws IOException, UnsupportedAudioFileException, PorcupineException {
        final String language = "es";
        final String[] keywords = {"murciélago"};
        porcupine = new Porcupine.Builder()
                .setAccessKey(accessKey)
                .setModelPath(PorcupineTestUtils.getTestModelPath(language))
                .setKeywordPaths(PorcupineTestUtils.getTestKeywordPaths(language, keywords))
                .build();

        runTestCase(
                "murciélago.wav",
                new ArrayList<>(Arrays.asList(0, 0)));
    }

    @ParameterizedTest(name = "testSingleKeyword for ''{0}''")
    @MethodSource("singleKeywordProvider")
    void testSingleKeyword(String language, String keyword, String audioFileName)
            throws IOException, UnsupportedAudioFileException, PorcupineException {
        final String[] keywords = {keyword};
        porcupine = new Porcupine.Builder()
                .setAccessKey(accessKey)
                .setModelPath(PorcupineTestUtils.getTestModelPath(language))
                .setKeywordPaths(PorcupineTestUtils.getTestKeywordPaths(language, keywords))
                .build();
        runTestCase(
                audioFileName,
                new ArrayList<>(Collections.singletonList(0)));
    }

    @ParameterizedTest(name = "testMultipleKeywords for ''{0}''")
    @MethodSource("multipleKeywordsProvider")
    void testSingleKeyword(
            String language,
            String[] keywords,
            String audioFileName,
            ArrayList<Integer> results)
            throws IOException, UnsupportedAudioFileException, PorcupineException {
        porcupine = new Porcupine.Builder()
                .setAccessKey(accessKey)
                .setModelPath(PorcupineTestUtils.getTestModelPath(language))
                .setKeywordPaths(PorcupineTestUtils.getTestKeywordPaths(language, keywords))
                .build();
        runTestCase(audioFileName, results);
    }
}

File: binding/go/porcupine_test.go
// Copyright 2021-2023 Picovoice Inc.
//
// You may not use this file except in compliance with the license. A copy of the license is
// located in the "LICENSE" file accompanying this source.
//
// Unless required by applicable law or agreed to in writing, software distributed under the
// License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied. See the License for the specific language governing permissions and
// limitations under the License.

package porcupine

import (
	"encoding/binary"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

var (
	testAccessKey                 string
	porcupine                     Porcupine
	singleKeywordTestParameters   []SingleKeywordTestData
	multipleKeywordTestParameters []MultipleKeywordTestData
)

type SingleKeywordTestData struct {
	language      string
	keyword       string
	testAudioFile string
}

type MultipleKeywordTestData struct {
	language      string
	keywords      []string
	testAudioFile string
	groundTruth   []int
}

func TestMain(m *testing.M) {

	flag.StringVar(&testAccessKey, "access_key", "", "AccessKey for testing")
	flag.Parse()

	singleKeywordTestParameters, multipleKeywordTestParameters = loadTestData()
	os.Exit(m.Run())
}

func appendLanguage(s string, language string) string {
	if language == "en" {
		return s
	}
	return s + "_" + language
}

func getTestModelPath(language string) string {
	modelRelPath := fmt.Sprintf(
		"../../lib/common/%s.pv",
		appendLanguage("porcupine_params", language))
	modelPath, _ := filepath.Abs(modelRelPath)
	return modelPath
}

func getTestKeywordPath(language string, keyword string) string {
	keywordRelPath := fmt.Sprintf(
		"../../resources/%s/%s/%s_%s.ppn",
		appendLanguage("keyword_files", language),
		osName,
		keyword,
		osName)
	keywordPath, _ := filepath.Abs(keywordRelPath)
	return keywordPath
}

func getTestKeywordPaths(language string, keywords []string) []string {
	keywordPaths := []string{}
	for _, keyword := range keywords {
		keywordPaths = append(keywordPaths, getTestKeywordPath(language, keyword))
	}
	return keywordPaths
}

func loadTestData() ([]SingleKeywordTestData, []MultipleKeywordTestData) {

	content, err := ioutil.ReadFile("../../resources/.test/test_data.json")
	if err != nil {
		log.Fatalf("Could not read test data json: %v", err)
	}

	var testData struct {
		Tests struct {
			SingleKeyword []struct {
				Language string `json:"language"`
				Keyword  string `json:"wakeword"`
			} `json:"singleKeyword"`
			MultipleKeyword []struct {
				Language    string   `json:"language"`
				Keywords    []string `json:"wakewords"`
				GroundTruth []int    `json:"groundTruth"`
			} `json:"multipleKeyword"`
		} `json:"tests"`
	}
	err = json.Unmarshal(content, &testData)
	if err != nil {
		log.Fatalf("Could not decode test data json: %v", err)
	}

	for _, x := range testData.Tests.SingleKeyword {
		singleKeywordTestData := SingleKeywordTestData{
			language:      x.Language,
			keyword:       x.Keyword,
			testAudioFile: strings.Replace(x.Keyword, " ", "_", -1) + ".wav",
		}

		singleKeywordTestParameters = append(singleKeywordTestParameters, singleKeywordTestData)
	}

	for _, x := range testData.Tests.MultipleKeyword {
		multipleKeywordTestData := MultipleKeywordTestData{
			language:      x.Language,
			keywords:      x.Keywords,
			testAudioFile: appendLanguage("multiple_keywords", x.Language) + ".wav",
			groundTruth:   x.GroundTruth,
		}

		multipleKeywordTestParameters = append(multipleKeywordTestParameters, multipleKeywordTestData)
	}

	return singleKeywordTestParameters, multipleKeywordTestParameters
}

func runTestCase(
	t *testing.T,
	porcupine *Porcupine,
	audioFileName string,
	expectedResults []int) {

	err := porcupine.Init()
	if err != nil {
		t.Fatalf("%v", err)
	}

	fmt.Printf("Porcupine Version: %s\n", Version)
	fmt.Printf("Frame Length: %d\n", FrameLength)
	fmt.Printf("Sample Rate: %d\n", SampleRate)

	testAudioPath, _ := filepath.Abs(filepath.Join("../../resources/audio_samples", audioFileName))
	data, err := ioutil.ReadFile(testAudioPath)
	if err != nil {
		t.Fatalf("Could not read test file: %v", err)
	}
	data = data[44:] // skip header

	var results []int
	frameLenBytes := FrameLength * 2
	frameCount := int(math.Floor(float64(len(data)) / float64(frameLenBytes)))
	sampleBuffer := make([]int16, FrameLength)
	for i := 0; i < frameCount; i++ {
		start := i * frameLenBytes

		for j := 0; j < FrameLength; j++ {
			dataOffset := start + (j * 2)
			sampleBuffer[j] = int16(binary.LittleEndian.Uint16(data[dataOffset : dataOffset+2]))
		}

		result, err := porcupine.Process(sampleBuffer)
		if err != nil {
			t.Fatalf("Could not read test file: %v", err)
		}
		if result >= 0 {
			fmt.Printf("Keyword %d triggered at %f\n", result, float64((i+1)*FrameLength)/float64(SampleRate))
			results = append(results, result)
		}
	}

	if !reflect.DeepEqual(results, expectedResults) {
		t.Fatalf("Detections are not correct")
	}
}

func TestSingleBuiltIn(t *testing.T) {

	porcupine = Porcupine{
		AccessKey:       testAccessKey,
		BuiltInKeywords: []BuiltInKeyword{PORCUPINE}}

	runTestCase(
		t,
		&porcupine,
		"porcupine.wav",
		[]int{0})

	delErr := porcupine.Delete()
	if delErr != nil {
		t.Fatalf("%v", delErr)
	}
}

func TestSingleKeyword(t *testing.T) {

	for _, tt := range singleKeywordTestParameters {
		t.Run(tt.language+" single keyword", func(t *testing.T) {
			porcupine = Porcupine{
				AccessKey:    testAccessKey,
				ModelPath:    getTestModelPath(tt.language),
				KeywordPaths: getTestKeywordPaths(tt.language, []string{tt.keyword})}
			runTestCase(
				t,
				&porcupine,
				tt.testAudioFile,
				[]int{0})
			delErr := porcupine.Delete()
			if delErr != nil {
				t.Fatalf("%v", delErr)
			}
		})
	}
}

func TestMultipleBuiltIns(t *testing.T) {

	porcupine = Porcupine{
		AccessKey: testAccessKey,
		BuiltInKeywords: []BuiltInKeyword{
			ALEXA,
			AMERICANO,
			BLUEBERRY,
			BUMBLEBEE,
			GRAPEFRUIT,
			GRASSHOPPER,
			PICOVOICE,
			PORCUPINE,
			TERMINATOR}}
	runTestCase(
		t,
		&porcupine,
		"multiple_keywords.wav",
		[]int{7, 0, 1, 2, 3, 4, 5, 6, 7, 8})
	delErr := porcupine.Delete()
	if delErr != nil {
		t.Fatalf("%v", delErr)
	}
}

func TestMultipleKeywords(t *testing.T) {
	for _, tt := range multipleKeywordTestParameters {
		t.Run(tt.language+" multiple keywords", func(t *testing.T) {
			porcupine = Porcupine{
				AccessKey:    testAccessKey,
				ModelPath:    getTestModelPath(tt.language),
				KeywordPaths: getTestKeywordPaths(tt.language, tt.keywords)}
			runTestCase(
				t,
				&porcupine,
				tt.testAudioFile,
				tt.groundTruth)
			delErr := porcupine.Delete()
			if delErr != nil {
				t.Fatalf("%v", delErr)
			}
		})
	}
}

func TestWithNonAsciiModelName(t *testing.T) {
	language := "es"
	keywords := []string{"murciélago"}
	porcupine = Porcupine{
		AccessKey:    testAccessKey,
		ModelPath:    getTestModelPath(language),
		KeywordPaths: getTestKeywordPaths(language, keywords)}
	runTestCase(
		t,
		&porcupine,
		"murciélago.wav",
		[]int{0, 0})
	delErr := porcupine.Delete()
	if delErr != nil {
		t.Fatalf("%v", delErr)
	}
}
